### PR TITLE
ON HOLD UNTIL MASTER/SLAVE REINTRODUCED:  remove hardcoded drop of openshift plugin; config.xml fixes; readme u…

### PR DESCRIPTION
**Pull request by gabemontero**
…pdates; update plugin version

@bparees @mfojtik these are various changes resulting from my working through the master/slave flow

Note: the ongoing dialogue with @ironcladlou regarding deployment trigger behavior could still affect the build steps in the 3 sample jobs, based on what happens with our use of the automatic flag, but i wanted to get a working version that still has a reasonable ci/cd workflow ready.

I've also have had to bump the plugin version based on the fixes (as well as the introduction of the env var changes).  Once it is available on the jenkins update center we can move forward with merging once the review is complete.

### CODE REVIEW

**Comment by gabemontero** [comment]
The exchange with @ironcladlou reached a conclusion.  I confirmed with @bparees my interpretation that for the short term at least, we'll leverage automatic=true for the deployment config's related to our master/slave sample.

A copy/paste of the relevant piece from @ironcladlou : 
So, I sublty misread the original report from Gabe and ALSO managed to
miss a subtle detail in the code, so let me try explaining this again.

The first (initial) deployment will be generated by resolving images
against any image triggers present regardless of the automatic flag,
by design. The deploymentConfig users create with image triggers don't
usually initially contain valid pullspecs. We resolve them once to get
the deployment started with the latest tag values defined in the
triggers. From that point forward, new image values for the tags will
*only* be resolved/updated if:

1. The referenced ImageStream tags change AND automatic=true, or
2. A configChange trigger is present and something *in the
deploymentConfig template changes* [1]

My note about the behavior of `oc deploy --latest` is still accurate:
images won't be resolved/updated in that case.

[1] This is what I think could be considered a bug; if there's a
template change that would normally trigger a new deployment, but the
image trigger is automatic=false, it might not make sense for images
to be resolved again.

File: examples/master/README.md [hunk 1]
```diff
@@ -68,18 +68,27 @@ plugins and configuration needed.
 
 ### Sample Application
 
-The last step is to instantiate the `sample-app` template. The [sample
-app](sample-app) is a simple Ruby application
-that runs Sinatra and has one unit test defined to exercise the CI flow.
+The last step is to instantiate the `sample-app` templates.  Look for the `sample-app-masterslave-ci.json` and
+`sample-app-masterslave-stage.json` files under the [jenkins example folder in OpenShift Origin](https://github.com/openshift/origin/tree/master/examples/jenkins).
+These two templates revolve around a [simple Ruby application that runs Sinatra](https://github.com/mfojtik/sample-app) and has one unit test defined to exercise the CI flow.
```
Comment (bparees) [inline examples/master/README.md]: ick.  we should really be pointing to https://github.com/openshift/ruby-hello-world as a sample instead (but we'd need to add unit testing to it)
@mfojtik

File: examples/master/README.md [hunk 2]
```diff
@@ -68,18 +68,27 @@ plugins and configuration needed.
...
 
-You have to instantiate the template in both `ci` and `stage` projects.
+Instantiate each template in their corresponding projects (`sample-app-masterslave-ci.json` in `ci` and `sample-app-masterslave-stage.json` in `stage`).
```
Comment (bparees) [inline examples/master/README.md]: so we should strip this down to just a regular jenkins, not master/slave.

File: examples/master/README.md [hunk 3]
```diff
@@ -68,18 +68,27 @@ plugins and configuration needed.
...
+video that shows the full workflow. The video includes one addition which we could not
+preconfigure in this sample (valid email accounts you have).
+
+But that detail aside, The breakdown of this sample workflow that is illustrated in the the video is:
 
-1. When the `sample-app-test` job is started it fetches the [sample-app](sample-app) sources,
+1. When the `sample-app-test` job is started it fetches the [sample-app](https://github.com/mfojtik/sample-app) sources,
    installs all required rubygems using bundler and then executes the sample unit tests.
    In the job definition, we restricted this job to run only on slaves that have
    the *ruby-20-centos7* label. This will match the Kubernetes Pod Template you see
```
Comment (bparees) [inline examples/master/README.md]: should update to centos/ruby-22-centos7 image also.

File: examples/master/README.md [hunk 4]
```diff
@@ -89,20 +98,30 @@ video that shows the full workflow. What happens in the video is:
...
-
-3. If the `sample-app` image passes the `stage` testing, you have to **manually
-   promote** the `sample-app-build` build to be deployed to OpenShift. Since
-   re-deploying the application replaces the existing running application, human
-   intervention is needed to confirm this step.
-
-4. Once the build is promoted, the `sample-app-deploy` job is started. This job
-   will scale down the existing application deployment and redeploy it using the
-   new Docker image.
+   `sample-app-stage` job will validate that the new Docker imageis successfully deploy with a running replic into the `stage` project.  The
```
Comment (bparees) [inline examples/master/README.md]: s/imageis/image is/
s/deploy/deployed/
s/replic/replica/

File: examples/master/README.md [hunk 5]
```diff
@@ -89,20 +98,30 @@ video that shows the full workflow. What happens in the video is:
...
+   deployment validation is started automatically because the `sample-app-stage` job continually monitors the ImageStream where the image resides
+   in the `ci` project and will detect any change in the ImageStream produced by the `sample-app-build` job.
+
+   A sample polling interval has been provided with the sample, but alter the interval to best suit your needs.  Also, no email notification is
+   preconfigured (a running SMTP server and a valid email are needed for the job to succeed), but you can set up SMTP in your Jenkins environment and
```
Comment (bparees) [inline examples/master/README.md]: s/job/step/ ?

File: examples/master/README.md [hunk 6]
```diff
@@ -89,20 +98,30 @@ video that shows the full workflow. What happens in the video is:
...
+   add a Jenkins email post build action to notify the QA team about your new image's availability for testing.
+   
+   Also note, the `sample-app-stage` job will start polling the `ci` project as soon as the Jenkins master comes up.  Jenkins' SCM polling infrastructure will 
+   trigger the job at least once because there are no existing job results, irregardless of the SCM commit ID polling information returned.  So you'll see
```
Comment (bparees) [inline examples/master/README.md]: s/irregardless/regardless/

Comment (bparees) [inline examples/master/README.md]: is this polling SCM or the imagestream?

Comment (bparees) [inline examples/master/README.md]: is this a jenkins limitation or our limitation?  doesn't sound like a great user experience.

File: examples/master/README.md [hunk 7]
```diff
@@ -89,20 +98,30 @@ video that shows the full workflow. What happens in the video is:
...
+   an initial run of `sample-app-stage` attempted and failed because you have not run `sample-app-test` yet, which would result in a `sample-app-build` run
+   that produces the deployment the `sample-app-stage` job is trying to validate.
+
+4. If the `sample-app` image passes whatever `stage` testing you deem appropriate, this sample workflow here introduces the human element of signing off
+   on the new Docker image.  You perform this sign off when you **manually promote** the `sample-app-build` build that served as input into the `stage` testing.
+   This provides to go ahead for the image to be spun up in the `ci` project and replace any existing, earlier versions of `sample-app` application.
+
+5. Once the build is promoted, the `sample-app-deploy` job is started. This job
+   will scale up the new application deployment (remember, the deployment in the `ci` project triggered by the new build is configured to 0 replicas to allow the testing
+   in to occur first), and then verify the new deployment is available with a running replica.
```
Comment (bparees) [inline examples/master/README.md]: s/in//

File: examples/master/configuration/jobs/sample-app-build/config.xml [hunk 8]
```diff
@@ -27,6 +27,13 @@
   <concurrentBuild>false</concurrentBuild>
   <builders>
 
+    <!--
+	the sample-app deployment config defined at https://github.com/openshift/origin/examples/jenkins/sample-app-masterslave-ci.json
+	sets replicas at 0, since we do not want to deploy until the testing in the stage project has occurred, and the promotion has been
+	manually approved.  Between the 2 steps here verifications below will no replicas are up, but that a new deployment has been triggered
```
Comment (bparees) [inline examples/master/configuration/jobs/sample-app-build/config.xml]: seems to be missing a word here.

File: examples/master/README.md (hunk 2 above)
Comment (gabemontero) [inline examples/master/README.md]: just making sure ... you mean sample-app-jenkins-ci.json and sample-app-jenkins-stage.json, correct? ... or maybe jenkins-ci.json and jenkins-stage.json ?

File: examples/master/README.md (hunk 5 above)
Comment (gabemontero) [inline examples/master/README.md]: it is a post build action actually ..

File: examples/master/README.md (hunk 6 above)
Comment (gabemontero) [inline examples/master/README.md]: It is our polling SCM implementation that polls OpenShift ImageStreams.  I'll try to be more explicit in saying that.

It is a jenkins limitation.  I tried to indicate that *nicely* but will update with something even more  explicit.  Agreed on ick factor.  The alternative, which @mfojtik had before, is to have a step which should always succeed, i.e. duplicate the build config and rebuild in the stage project, which as we discussed did not seem ideal either.

File: examples/master/configuration/jobs/sample-app-build/config.xml (hunk 8 above)
Comment (gabemontero) [inline examples/master/configuration/jobs/sample-app-build/config.xml]: yep ... rewording is coming ...

**Comment by gabemontero** [comment]
updates pushed for the various typos and rewording form @bparees 's comment.  I'll rename the json files pending the answer to my request for clarification.

File: examples/master/README.md (hunk 2 above)
Comment (bparees) [inline examples/master/README.md]: yes, i mean the templates (and their names) shouldn't do any slave related stuff.

File: examples/master/README.md (hunk 1 above)
Comment (mfojtik) [inline examples/master/README.md]: @bparees +1

I added unit tests to that app just for this reason. @gabemontero can you please change it?

**Comment by mfojtik** [comment]
LGTM

**Comment by mfojtik** [comment]
@gabemontero @bparees still valid PR? Should have at least less horrifying title :-)

**Comment by gabemontero** [comment]
@mfojtik @bparees the files affected are not present in @mfojtik 's latest pull https://github.com/openshift/jenkins/pull/77 in the slave space .... agreed, this pull is no longer relevant.

Closing out.

**Pull request closed without merging**